be/src/exec/stream/aggregate/stream_aggregator.cpp

```diff
@@ -0,0 +1,585 @@
+    VLOG_ROW << " output post ops:"
+             << ", debug_string:" << post_ops->debug_string();
+#endif
+
+    auto post_chunk_result =
+            _build_output_chunk_with_ops(post_group_by_columns, post_agg_result_columns, std::move(post_ops), false);
+    DCHECK_EQ(selected_count, post_chunk_result->num_rows());
+
+    prev_result_chunk->append_safe(*post_chunk_result);
+    auto* pre_result_chunk_ops = dynamic_cast<ChunkExtraColumnsData*>(prev_result_chunk->get_extra_data().get());
+    auto* post_result_chunk_ops = dynamic_cast<ChunkExtraColumnsData*>(post_chunk_result->get_extra_data().get());
+    pre_result_chunk_ops->append(*post_result_chunk_ops, 0, post_result_chunk_ops->num_rows());
+
+    *result_chunk = prev_result_chunk;
+
+    return Status::OK();
+}
+
+Status StreamAggregator::_output_result_prev_retract_changes(size_t chunk_size,
+                                                             const vectorized::Columns& group_by_columns,
+                                                             const vectorized::Column* count_column,
+                                                             vectorized::StreamChunkPtr* prev_result_chunk,
+                                                             vectorized::Buffer<uint8_t>* prev_existence) {
+    vectorized::Columns prev_group_by_columns = _create_group_by_columns(chunk_size);
+    vectorized::Columns prev_agg_result_columns = _create_agg_result_columns(chunk_size, false);
+    Int8ColumnPtr prev_ops = vectorized::Int8Column::create();
+
+    std::vector<DatumRow> keys;
+    keys.reserve(chunk_size);
+    for (size_t i = 0; i < chunk_size; i++) {
+        keys.emplace_back(_convert_to_datum_row(group_by_columns, i));
+    }
+    auto result_ors = _result_state_table->seek(keys);
+    DCHECK_EQ(result_ors.size(), chunk_size);
+
+    // NOTE: Always retract old agg results first.
+    // Should we all keep UPDATE_BEFOER/UDPATE_AFTER serially?
+    vectorized::Buffer<uint32_t> permutation;
+    prev_existence->assign(chunk_size, 0);
+
+    auto agg_count_column = reinterpret_cast<const vectorized::Int64Column*>(count_column);
+    auto agg_count_column_data = agg_count_column->get_data();
+    for (size_t i = 0; i < chunk_size; i++) {
+        auto result = result_ors[i];
+        if (!result.ok()) continue;
+        auto prev_result = result.value();
+        DCHECK_EQ(prev_result->num_rows(), 1);
+        // TODO: prev_result_table should keep ops column?
+        DCHECK_LE(prev_agg_result_columns.size(), prev_result->num_columns());
+
+        // append agg columns
+        for (size_t idx = 0; idx < prev_agg_result_columns.size(); idx++) {
+            auto& prev_col = prev_result->get_column_by_index(idx);
+            prev_agg_result_columns[idx]->append(*prev_col);
+        }
+        permutation.push_back(i);
+        (*prev_existence)[i] = 1;
+
+        // generate op
+        if (agg_count_column_data[i] > 0) {
+            prev_ops->append(UPDATE_BEFORE_OP);
```

Comment:
How to make sure there's an `UPDATE_AFTER` following it?

Reply:
For vectorized handling, only can ensure  `UPDATE_BEFORE` and `UPDATE_AFTER ` are in the same chunk. and `UPDATE_BEFORE` are before  `UPDATE_AFTER`.

I doubt whether this can be satisfied without exactly just following.  Maybe change it latter it there is some requirements.